### DISCUSSION TITLE
`animate background` not triggering as expected

### DISCUSSION THREAD

**Discussion by mitchgrout** [slint-ui/slint#4807] [Q&A] [answered]
Consider the following fragment:

```js
export component MainWindow inherits Window {
  private property <duration> duration:    1s;
  private property <[color]>  colors:      [#f00, #0f0, #00f];
  private property <int>      tick:        animation-tick() / duration;
  private property <int>      color-index: mod(tick, colors.length);

  width:      400px;
  height:     300px;
  background: colors[color-index];

  animate background { duration: self.duration; }
}
```

The behaviour I expect is that over a 3s period, the window will cycle smoothly between red, green, and blue, then repeat. On v1.4.1 (VSCode extension + SlintPad), the window instead remains at _one_ of the colours, and will not change.

If I tack on something like...

```js
  TouchArea {
    pointer-event(e) => {
      if (e.kind == PointerEventKind.up) { parent.tick += 1; }
    }
  }
```

...then once I click the area, the `background` will finally update to the next colour in the cycle. Observing the values of `tick` and `color-index`, they are changing as expected, but the animation does not appear to be kicking in. Thoughts?

**Comment by mitchgrout** [comment]
Hmm, I have an answer but it leads to another question. I took my original fragment, and added a `Text` component to monitor the values of `tick` and `color-index`. These update on the screen as expected, without the color changing. However, if I jiggle my mouse over the screen, then the animation kicks in. I'm a bit surprised that the text can re-render itself without invalidating at least a portion of its parent? Or perhaps, the `animate background` is not being triggered without some sort of user event to force it?

**Reply by mitchgrout** [reply]
I've set this up in [SlintPad](https://slintpad.com/?snippet=export+component+MainWindow+inherits+Window+%7B%0A++private+property+%3Cduration%3E+duration%3A++++1s%3B%0A++private+property+%3C%5Bcolor%5D%3E++colors%3A++++++%5B%23f00%2C+%230f0%2C+%2300f%5D%3B%0A++private+property+%3Cint%3E++++++tick%3A++++++++animation-tick%28%29+%2F+duration%3B%0A++private+property+%3Cint%3E++++++color-index%3A+mod%28tick%2C+colors.length%29%3B%0A%0A++width%3A++++++400px%3B%0A++height%3A+++++300px%3B%0A++background%3A+colors%5Bcolor-index%5D%3B%0A%0A++animate+background+%7B+duration%3A+100ms%3B+%7D%0A%0A++%2F%2F+TouchArea+%7B%0A++%2F%2F+++pointer-event%28e%29+%3D%3E+%7B%0A++%2F%2F+++++if+%28e.kind+%3D%3D+PointerEventKind.up%29+%7B+parent.value+%2B%3D+1%3B+%7D%0A++%2F%2F+++%7D%0A++%2F%2F+%7D%0A%0A++Text+%7B%0A++++text%3A+%22tick+%3D+%22+%2B+tick+%2B+%22%3B+color-index+%3D+%22+%2B+color-index%3B%0A++%7D%0A%7D%0A&style=native) for ease of access. As noted above:

- Once loaded, will render and remain on one of the colours
- Text area will continue to update and re-render
- Moving the mouse over the window will cause the animation to begin to trigger

**Comment by hunger** [comment]
~~#4752 fixed a bug with animations not running as expected recently, but that seems unrelated to me.~~

**Comment by ogoffart** [comment, marked as answer]
I have seen this problem before, but there was not yet an open bug about it so i opened #4811

The problem is that because of animation-tick, the property is always marked as dirty and the animation is constantly restarted without having the ability to do any progress.
